### DISCUSSION TITLE
How to write a space-like four-vector in FeynCalc?

### DISCUSSION THREAD

**Discussion by alessiotbceleste** [FeynCalc/feyncalc#222] [General]
Hi, please ...

How to write a space-like quadrivector in FeynCalc in Mathematica?

**Comment by vsht** [comment]
Hi,

sorry, but I don't really understand what you are trying to achieve?

**Comment by alessiotbceleste** [comment]
for example, in FeynCalc the command TC[p] is the temporal component of a 4-vector, pt = (p0, 0, 0, 0).  Is there a similar one for the spatial component? ps = (0, p1, p2, p3)

**Comment by vsht** [comment]
You mean `CV[p,i]`?

https://feyncalc.github.io/FeynCalcBookDev/CV.html

**Comment by alessiotbceleste** [comment]
hi thanks for the tips, I had thought of using this command but it would have 3 components, I was thinking a four-vector with 4 components, where the temporal part is equal to zero.
Would this be possible?

**Comment by vsht** [comment]
This is why I asked what are you trying to achieve in the first place. For a Lorentz covariant calculation it should be sufficient to satisfy
all such constraints by just assigning dedicated values to the corresponding via scalar products.

How do you expect your vector to behave differently from a usual `FV`-vector?

**Comment by alessiotbceleste** [comment]
the calculation I'm trying to implement is the trace below, where vector b is like space

bt = TemporalPair[TemporalMomentum[b], ExplicitLorentzIndex[0]];

b_alpha1 = Contract[FV[b, \alpha_1] - bt];
b_alpha2 = Contract[FV[b, \alpha_2]] - bt];

LINE = GS[p_2] . b_alpha1 . GA[\beta_1] . FV[p_1 - p_3, \rho_1] . LC[\alpha_1, \beta_1, \rho_1, \lambda_1] . MT[\lambda_1, \mu] . GS[p_1 - p_3] .
            b_alpha2 . GA[\beta_2] . FV[p_1 - p_3, \rho_2] . LC[\alpha_2, \beta_2, \rho_2, \lambda_2] . MT[\lambda_2, \nu] . GS[p_1] . GA[\nu] . 
            GS[p_1 - p_3] . GA[\mu]

Contract[Tr[LINE]]

**Comment by vsht** [comment]
Try to paste your code back into a Mathematica notebook: it doesn't get interpreted correctly. Please use Copy as -> Input text.

**Reply by alessiotbceleste** [reply]
bt=TemporalPair[TemporalMomentum[b],ExplicitLorentzIndex[0]];

balpha1=Contract[FV[b,Subscript[\[Alpha], 1]]-bt];

balpha2=Contract[FV[b,Subscript[\[Alpha], 2]]-bt];

LINE=prop[Subscript[p, 2],-m].(-balpha1.GA[Subscript[\[Beta], 1]].FV[Subscript[p, 1]-Subscript[p, 3],Subscript[\[Rho], 1]].LC[Subscript[\[Alpha], 1],Subscript[\[Beta], 1],Subscript[\[Rho], 1],Subscript[\[Lambda], 1]].MT[Subscript[\[Lambda], 1],\[Mu]]).prop[Subscript[p, 1]-Subscript[p, 3],m].(-balpha2.GA[Subscript[\[Beta], 2]].FV[Subscript[p, 1]-Subscript[p, 3],Subscript[\[Rho], 2]].LC[Subscript[\[Alpha], 2],Subscript[\[Beta], 2],Subscript[\[Rho], 2],Subscript[\[Lambda], 2]].MT[Subscript[\[Lambda], 2],\[Nu]]).prop[Subscript[p, 1],m].GA[\[Nu]].prop[Subscript[p, 1]-Subscript[p, 3],m].GA[\[Mu]];

M = Contract[Tr[LINE]];

**Reply by vsht** [reply]
Thanks. Does this help?

```
FCClearScalarProducts[]
Pair[Momentum[b], Momentum[x_]] := FCI[TC[b] TC[x]]
M = Contract[DiracTrace[LINE]]
```

```
M // DiracSimplify // Simplify
```

```
-16*prop[Subscript[p, 1], m]*prop[Subscript[p, 2], -m]*prop[Subscript[p, 1] - Subscript[p, 3], m]^2*TC[b]*
 (-(FV[b, Subscript[\[Alpha], 1]]*(SP[Subscript[p, 1], Subscript[p, 1]] - 2*SP[Subscript[p, 1], Subscript[p, 3]] + SP[Subscript[p, 3], Subscript[p, 3]])) - 
  FV[b, Subscript[\[Alpha], 2]]*(SP[Subscript[p, 1], Subscript[p, 1]] - 2*SP[Subscript[p, 1], Subscript[p, 3]] + SP[Subscript[p, 3], Subscript[p, 3]]) + 
  TC[b]*(SP[Subscript[p, 1], Subscript[p, 1]] + MT[Subscript[\[Alpha], 1], Subscript[\[Alpha], 2]]*SP[Subscript[p, 1], Subscript[p, 1]] - 2*SP[Subscript[p, 1], Subscript[p, 3]] - 
    2*MT[Subscript[\[Alpha], 1], Subscript[\[Alpha], 2]]*SP[Subscript[p, 1], Subscript[p, 3]] + SP[Subscript[p, 3], Subscript[p, 3]] + MT[Subscript[\[Alpha], 1], Subscript[\[Alpha], 2]]*SP[Subscript[p, 3], Subscript[p, 3]] + 
    FV[Subscript[p, 1], Subscript[\[Alpha], 2]]*TC[Subscript[p, 1]] - FV[Subscript[p, 3], Subscript[\[Alpha], 2]]*TC[Subscript[p, 1]] - TC[Subscript[p, 1]]^2 + 
    FV[Subscript[p, 1], Subscript[\[Alpha], 1]]*(-FV[Subscript[p, 1], Subscript[\[Alpha], 2]] + FV[Subscript[p, 3], Subscript[\[Alpha], 2]] + TC[Subscript[p, 1]] - TC[Subscript[p, 3]]) - 
    FV[Subscript[p, 1], Subscript[\[Alpha], 2]]*TC[Subscript[p, 3]] + FV[Subscript[p, 3], Subscript[\[Alpha], 2]]*TC[Subscript[p, 3]] + 2*TC[Subscript[p, 1]]*TC[Subscript[p, 3]] - TC[Subscript[p, 3]]^2 + 
    FV[Subscript[p, 3], Subscript[\[Alpha], 1]]*(FV[Subscript[p, 1], Subscript[\[Alpha], 2]] - FV[Subscript[p, 3], Subscript[\[Alpha], 2]] - TC[Subscript[p, 1]] + TC[Subscript[p, 3]])))
```

**Reply by alessiotbceleste** [reply]
Thank you very much, although I didn't understand how these command lines work, I'll test it here. 
I forgot to inform you that the command prop[p_,m_] = GS[p]+m

**Comment by vsht** [comment]
The command just defines that every scalar product of the form b.x should be replaced with b^0 x^0, which effectively implements the constraint that the spatial part of b is zero.


Am 12. Juni 2023 02:01:43 UTC schrieb alessiotbceleste ***@***.***>:

**Comment by alessiotbceleste** [comment]
Hi, first thank you so much for the comments, but what I intended was that the temporal part of b is zero and the spatial one is not zero. Follow my code again ...

<<FeynCalc`

m=0;
prop[p_,m_]:=GS[p]+m

bt=TemporalPair[TemporalMomentum[b],ExplicitLorentzIndex[0]];

balpha1=Contract[FV[b,Subscript[\[Alpha], 1]]-bt]; 
balpha2=Contract[FV[b,Subscript[\[Alpha], 2]]-bt];
balpha3=Contract[FV[b,Subscript[\[Alpha], 3]]-bt];
balpha4=Contract[FV[b,Subscript[\[Alpha], 4]]-bt];

LINE=prop[Subscript[p, 2],-m].(-balpha1.GA[Subscript[\[Beta], 1]].FV[Subscript[p, 1]-Subscript[p, 3],Subscript[\[Rho], 1]].LC[Subscript[\[Alpha], 1],Subscript[\[Beta], 1],Subscript[\[Rho], 1],Subscript[\[Lambda], 1]].MT[Subscript[\[Lambda], 1],\[Mu]]).prop[Subscript[p, 1]-Subscript[p, 3],m].(-balpha2.GA[Subscript[\[Beta], 2]].FV[Subscript[p, 1]-Subscript[p, 3],Subscript[\[Rho], 2]].LC[Subscript[\[Alpha], 2],Subscript[\[Beta], 2],Subscript[\[Rho], 2],Subscript[\[Lambda], 2]].MT[Subscript[\[Lambda], 2],\[Nu]]).prop[Subscript[p, 1],m].GA[\[Nu]].prop[Subscript[p, 1]-Subscript[p, 3],m].GA[\[Mu]];

M=Contract[Tr[LINE]];

**Reply by vsht** [reply]
Ok, then you can just write
```
FCClearScalarProducts[]
TC[b] = 0;
Pair[Momentum[b], Momentum[x_]] := FCI[CSP[b, x]];
M = DiracSimplify[DiracTrace[LINE]]
```

**Reply by alessiotbceleste** [reply]
Can I use LINE the way I wrote it myself?

**Reply by vsht** [reply]
In FeynCalc 10 `Tr` will not invoke `DiracTrace` anymore. It was a rather ugly hack that involve redefining system functions of Mathematica (`Tr`) so for the next release we want to get away from that.

**Comment by alessiotbceleste** [comment]
Hello! Could you help me with this code, because Mathematica doesn't perform the contraction?

*********************************************************************

FCClearScalarProducts[]
TC[b] = 0;
Pair[Momentum[b], Momentum[x_]] := FCI[CSP[b, x]];
bt = TemporalPair[TemporalMomentum[b], ExplicitLorentzIndex[0]];
bmu = FV[b, \[Mu]] - bt;
bnu = FV[b, \[Nu]] - bt;

line1 = GS[Subscript[p, 1]].(-FV[Subscript[p, 1] + Subscript[p, 2], \[Beta]].bmu.MT[\[Mu], \[Beta]] - SP[bmu, Subscript[p, 1] + Subscript[p, 2]].GA[\[Mu]]).GS[Subscript[p, 2]].GA[\[Nu]];

trace1 = DiracTrace[line1, DiracTraceEvaluate -> True];

*********************************************************************

Message produced by Mathematica 

Contract: Error! Contract has encountered a fatal problem and must abort the computation. The problem reads: Something went wrong during prepareProductContractions.

**Reply by vsht** [reply]
Hi,

the expression `SP[bmu, Subscript[p, 1] + Subscript[p, 2]]` makes no sense to me. The same goes for the definitions of `bmu` and `bnu` for that matter.

Still, the evaluation works for me, even though the result is not meaningful:

```
FCClearScalarProducts[]
TC[b] = 0;
Pair[Momentum[b], Momentum[x_]] := FCI[CSP[b, x]];
bt = TemporalPair[TemporalMomentum[b], ExplicitLorentzIndex[0]];
bmu = FV[b, \[Mu]] - bt;
bnu = FV[b, \[Nu]] - bt;
line1 = GS[
    Subscript[p, 
     1]] . (-FV[Subscript[p, 1] + Subscript[p, 2], \[Beta]] . bmu . 
       MT[\[Mu], \[Beta]] - 
     SP[bmu, Subscript[p, 1] + Subscript[p, 2]] GA[\[Mu]]) . 
   GS[Subscript[p, 2]] . GA[\[Nu]];
trace1 = DiracTrace[line1, DiracTraceEvaluate -> True];
```

trace1 // StandardForm
(*
-4 (Pair[LorentzIndex[\[Nu]], Momentum[Subscript[p, 2]]] Pair[
     Momentum[
      CartesianPair[CartesianMomentum[b], 
       CartesianMomentum[Subscript[p, 1]]]], 
     Momentum[Subscript[p, 1]]] + 
   Pair[LorentzIndex[\[Nu]], Momentum[Subscript[p, 2]]] Pair[
     Momentum[
      CartesianPair[CartesianMomentum[b], 
       CartesianMomentum[Subscript[p, 1]]]], 
     Momentum[Subscript[p, 2]]] + 
   Pair[LorentzIndex[\[Nu]], Momentum[Subscript[p, 1]]] Pair[
     Momentum[
      CartesianPair[CartesianMomentum[b], 
       CartesianMomentum[Subscript[p, 2]]]], 
     Momentum[Subscript[p, 1]]] + 
   Pair[LorentzIndex[\[Nu]], Momentum[Subscript[p, 1]]] Pair[
     Momentum[
      CartesianPair[CartesianMomentum[b], 
       CartesianMomentum[Subscript[p, 2]]]], 
     Momentum[Subscript[p, 2]]] - 
   Pair[Momentum[Pair[LorentzIndex[\[Nu]], Momentum[b]]], 
     Momentum[Subscript[p, 1]]] Pair[Momentum[Subscript[p, 1]], 
     Momentum[Subscript[p, 2]]] - 
   Pair[Momentum[Pair[LorentzIndex[\[Nu]], Momentum[b]]], 
     Momentum[Subscript[p, 2]]] Pair[Momentum[Subscript[p, 1]], 
     Momentum[Subscript[p, 2]]])
*)

**Comment by alessiotbceleste** [comment]
Hello, thanks for the help. I can't get a correct result for the dash calculation, here's the code used

FCClearScalarProducts[]
TC[b]=0;
Pair[Momentum[b],Momentum[x_]]:=FCI[CSP[b,x]];

bt=TemporalPair[TemporalMomentum[b],ExplicitLorentzIndex[0]];
bmu=FV[b,\[Mu]]-bt;

line=GS[p].(GS[q].GA[5].bmu-SP[bmu,q].GA[5].GA[\[Mu]]).GS[k].GA[\[Nu]];
trace=DiracTrace[line,DiracTraceEvaluate->True]

**Comment by vsht** [comment]
Hi,

what is `bmu` supposed to mean? We do agree that writing $(p^\mu - p^0)$ inside a Lorentz covariant expression looks very weird, right?

Then, I also don't understand what are you trying to express with `SP[bmu, q]`? Could you please write the corresponding expressions in TeX?

**Reply by alessiotbceleste** [reply]
Hi, answering your questions:

1. What should bmu mean?
Answer: I want b to be a space-like four vector.

2. What are you trying to express with SP[bmu, q]?
Answer: It would be the scalar product with the four-vector b with a four-momentum q.

$b \cdot q$

**Comment by vsht** [comment]
Hi,

then you can just subtract a time-like vector n^mu
from the full vector. In Lorentz covariant expressions
the fact that n^mu = (1,0,0,0) will be relevant only for
n^2 = 1.





Am 16.04.24 um 13:44 schrieb alessiotbceleste:
